File: src/packages/data/anomaly_detection/sdk/examples/javascript/streaming_detection.js
/**
 * Streaming Anomaly Detection Example using JavaScript SDK
 * 
 * This example demonstrates real-time anomaly detection using WebSocket streaming.
 */

const { 
    StreamingClient, 
    AlgorithmType,
    StreamingError 
} = require('@anomaly-detection/sdk');

const readline = require('readline');

/**
 * Example streaming anomaly detection class
 */
class AnomalyDetectionStream {
    constructor(wsUrl = 'ws://localhost:8000/ws/stream') {
        this.wsUrl = wsUrl;
        this.anomalyCount = 0;
        this.totalPoints = 0;
        this.running = false;
        
        // Configure streaming
        const config = {
            wsUrl: wsUrl,
            bufferSize: 50,
            detectionThreshold: 0.6,
            batchSize: 5,
            algorithm: AlgorithmType.ISOLATION_FOREST,
            autoRetrain: false,
            autoReconnect: true,
            reconnectDelay: 3000
        };
        
        // Initialize streaming client
        this.client = new StreamingClient(config);
        this.setupHandlers();
    }
    
    setupHandlers() {
        // Connection handlers
        this.client.on('connect', () => {
            console.log('✅ Connected to streaming service');
            this.running = true;
        });
        
        this.client.on('disconnect', () => {
            console.log('❌ Disconnected from streaming service');
            this.running = false;
        });
        
        // Anomaly detection handler
        this.client.on('anomaly', (anomalyData) => {
            this.anomalyCount++;
            console.log('🚨 ANOMALY DETECTED:');
            console.log(`   Index: ${anomalyData.index}`);
            console.log(`   Score: ${anomalyData.score.toFixed(4)}`);
            console.log(`   Data Point: [${anomalyData.dataPoint.map(x => x.toFixed(3)).join(', ')}]`);
            if (anomalyData.confidence) {
                console.log(`   Confidence: ${anomalyData.confidence.toFixed(4)}`);
            }
            console.log(`   Total anomalies so far: ${this.anomalyCount}`);
            console.log('-'.repeat(50));
        });
        
        // Error handler
        this.client.on('error', (error) => {
            console.error('❌ Streaming error:', error.message);
        });
        
        // Raw message handler (optional)
        this.client.on('message', (data) => {
            // Handle any other message types
            if (data.type !== 'anomaly' && data.type !== 'ping') {
                console.log('📨 Received message:', data);
            }
        });
    }
    
    async startStreaming() {
        console.log('Starting streaming anomaly detection...');
        await this.client.start();
    }
    
    stopStreaming() {
        console.log('Stopping streaming...');
        this.client.stop();
        this.running = false;
    }
    
    sendDataPoint(dataPoint) {
        try {
            this.client.sendData(dataPoint);
            this.totalPoints++;
        } catch (error) {
            console.error('Error sending data:', error.message);
        }
    }
    
    async generateAndSendData(duration = 30000, interval = 1000) {
        console.log(`Generating sample data for ${duration/1000} seconds...`);
        console.log('Normal data will be around [0, 0], anomalies around [5, 5]');
        console.log('-'.repeat(50));
        
        const startTime = Date.now();
        let intervalId;
        
        return new Promise((resolve) => {
            intervalId = setInterval(() => {
                if (Date.now() - startTime >= duration || !this.running) {
                    clearInterval(intervalId);
                    console.log(`\n✅ Sent ${this.totalPoints} data points`);
                    console.log(`✅ Detected ${this.anomalyCount} anomalies`);
                    resolve();
                    return;
                }
                
                // Generate mostly normal data with occasional anomalies
                let dataPoint;
                if (Math.random() < 0.1) { // 10% chance of anomaly
                    // Anomalous data point
                    dataPoint = [
                        Math.random() * 1 + 4.5, // Between 4.5 and 5.5
                        Math.random() * 1 + 4.5
                    ];
                    console.log(`📤 Sending anomalous point: [${dataPoint.map(x => x.toFixed(3)).join(', ')}]`);
                } else {
                    // Normal data point
                    dataPoint = [
                        (Math.random() - 0.5) * 2, // Between -1 and 1
                        (Math.random() - 0.5) * 2
                    ];
                    console.log(`📤 Sending normal point: [${dataPoint.map(x => x.toFixed(3)).join(', ')}]`);
                }
                
                this.sendDataPoint(dataPoint);
            }, interval);
        });
    }
}

/**
 * Interactive streaming example
 */
async function interactiveStreamingExample() {
    console.log('=== Interactive Streaming Example ===');
    
    const stream = new AnomalyDetectionStream();
    await stream.startStreaming();
    
    // Wait for connection
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    if (!stream.running) {
        console.log('Failed to connect to streaming service');
        return;
    }
    
    console.log('\nInteractive mode - Enter data points as comma-separated values');
    console.log('Example: 1.5,2.3 or 5.0,5.0 (for anomaly)');
    console.log('Type "quit" to exit');
    
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    const askForInput = () => {
        rl.question('\nEnter data point (x,y): ', (input) => {
            const trimmed = input.trim();
            
            if (trimmed.toLowerCase() === 'quit' || trimmed.toLowerCase() === 'exit' || trimmed === 'q') {
                rl.close();
                stream.stopStreaming();
                console.log(`\nSession summary:`);
                console.log(`Total points sent: ${stream.totalPoints}`);
                console.log(`Anomalies detected: ${stream.anomalyCount}`);
                return;
            }
            
            try {
                const values = trimmed.split(',').map(x => parseFloat(x.trim()));
                if (values.length !== 2 || values.some(isNaN)) {
                    console.log('Please enter exactly 2 numeric values separated by comma');
                } else {
                    stream.sendDataPoint(values);
                    console.log(`Sent: [${values.join(', ')}]`);
                }
            } catch (error) {
                console.log('Invalid input. Please enter numeric values separated by comma');
            }
            
            // Continue asking for input
            if (stream.running) {
                askForInput();
            }
        });
    };
    
    askForInput();
}

/**
 * Automated streaming example
 */
async function automatedStreamingExample() {
    console.log('=== Automated Streaming Example ===');
    
    const stream = new AnomalyDetectionStream();
    await stream.startStreaming();
    
    // Wait for connection
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    if (!stream.running) {
        console.log('Failed to connect to streaming service');
        return;
    }
    
    try {
        // Run automated data generation
        await stream.generateAndSendData(20000, 500); // 20 seconds, 500ms intervals
    } catch (error) {
        console.error('Automated streaming failed:', error);
    } finally {
        stream.stopStreaming();
    }
}

/**
 * Batch streaming example
 */
async function batchStreamingExample() {
    console.log('=== Batch Streaming Example ===');
    
    // Generate batch data
    const normalBatch = [];
    const anomalyBatch = [];
    
    for (let i = 0; i < 20; i++) {
        normalBatch.push([
            (Math.random() - 0.5) * 2,
            (Math.random() - 0.5) * 2
        ]);
    }
    
    for (let i = 0; i < 5; i++) {
        anomalyBatch.push([
            Math.random() * 1 + 4.5,
            Math.random() * 1 + 4.5
        ]);
    }
    
    const stream = new AnomalyDetectionStream();
    await stream.startStreaming();
    
    // Wait for connection
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    if (!stream.running) {
        console.log('Failed to connect to streaming service');
        return;
    }
    
    try {
        console.log('Sending normal data batch...');
        for (let i = 0; i < normalBatch.length; i++) {
            const point = normalBatch[i];
            stream.sendDataPoint(point);
            console.log(`Sent normal point ${i+1}/20: [${point.map(x => x.toFixed(3)).join(', ')}]`);
            await new Promise(resolve => setTimeout(resolve, 200));
        }
        
        console.log('\nSending anomalous data batch...');
        for (let i = 0; i < anomalyBatch.length; i++) {
            const point = anomalyBatch[i];
            stream.sendDataPoint(point);
            console.log(`Sent anomaly point ${i+1}/5: [${point.map(x => x.toFixed(3)).join(', ')}]`);
            await new Promise(resolve => setTimeout(resolve, 200));
        }
        
        // Wait for processing
        console.log('\nWaiting for processing to complete...');
        await new Promise(resolve => setTimeout(resolve, 5000));
        
    } catch (error) {
        console.error('Batch streaming failed:', error);
    } finally {
        stream.stopStreaming();
    }
}

/**
 * Performance streaming example
 */
async function performanceStreamingExample() {
    console.log('=== Performance Streaming Example ===');
    
    // Configure for high throughput
    const config = {
        wsUrl: 'ws://localhost:8000/ws/stream',
        bufferSize: 100,
        detectionThreshold: 0.5,
        batchSize: 20, // Larger batches for efficiency
        algorithm: AlgorithmType.ISOLATION_FOREST,
        autoReconnect: true
    };
    
    const client = new StreamingClient(config);
    
    // Performance tracking
    let anomalyCount = 0;
    let startTime = null;
    let pointsSent = 0;
    
    client.on('connect', () => {
        console.log('✅ Connected - Starting performance test');
        startTime = Date.now();
    });
    
    client.on('anomaly', (anomalyData) => {
        anomalyCount++;
        if (anomalyCount % 10 === 0) {
            const elapsed = (Date.now() - startTime) / 1000;
            const rate = elapsed > 0 ? pointsSent / elapsed : 0;
            console.log(`📊 Performance: ${anomalyCount} anomalies detected, ${rate.toFixed(1)} points/sec`);
        }
    });
    
    client.on('error', (error) => {
        console.error('❌ Error:', error.message);
    });
    
    try {
        await client.start();
        await new Promise(resolve => setTimeout(resolve, 2000)); // Wait for connection
        
        // Generate high volume of data
        console.log('Sending high volume data stream (1000 points)...');
        
        for (let i = 0; i < 1000; i++) {
            if (i % 100 === 0) {
                console.log(`Sent ${i}/1000 points...`);
            }
            
            // Mix of normal and anomalous data
            let dataPoint;
            if (Math.random() < 0.05) { // 5% anomalies
                dataPoint = [
                    Math.random() * 1 + 3.5,
                    Math.random() * 1 + 3.5
                ];
            } else {
                dataPoint = [
                    (Math.random() - 0.5) * 2,
                    (Math.random() - 0.5) * 2
                ];
            }
            
            client.sendData(dataPoint);
            pointsSent++;
            
            // Small delay to avoid overwhelming
            await new Promise(resolve => setTimeout(resolve, 10));
        }
        
        // Wait for final processing
        console.log('Waiting for final processing...');
        await new Promise(resolve => setTimeout(resolve, 5000));
        
        // Calculate final statistics
        const totalTime = (Date.now() - startTime) / 1000;
        const throughput = pointsSent / totalTime;
        
        console.log('\n📊 Performance Results:');
        console.log(`   Total points: ${pointsSent}`);
        console.log(`   Total time: ${totalTime.toFixed(2)} seconds`);
        console.log(`   Throughput: ${throughput.toFixed(1)} points/second`);
        console.log(`   Anomalies detected: ${anomalyCount}`);
        console.log(`   Anomaly rate: ${((anomalyCount/pointsSent)*100).toFixed(2)}%`);
        
    } catch (error) {
        console.error('Performance test failed:', error);
    } finally {
        client.stop();
    }
}

/**
 * Error handling example for streaming
 */
async function streamingErrorHandlingExample() {
    console.log('=== Streaming Error Handling Example ===');
    
    // Test with invalid URL
    const invalidClient = new StreamingClient({
        wsUrl: 'ws://invalid-url:9999/ws/stream',
        autoReconnect: false
    });
    
    let errorReceived = false;
    
    invalidClient.on('error', (error) => {
        console.log('✅ Expected connection error:', error.message);
        errorReceived = true;
    });
    
    try {
        await invalidClient.start();
        await new Promise(resolve => setTimeout(resolve, 5000));
        
        if (!errorReceived) {
            console.log('❌ Expected error but none received');
        }
        
    } catch (error) {
        console.log('✅ Connection failed as expected:', error.message);
    } finally {
        invalidClient.stop();
    }
    
    // Test sending invalid data
    console.log('\nTesting invalid data handling...');
    const validClient = new StreamingClient({
        wsUrl: 'ws://localhost:8000/ws/stream'
    });
    
    try {
        await validClient.start();
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Try to send invalid data
        try {
            validClient.sendData('invalid data');
            console.log('❌ Expected validation error but got success');
        } catch (error) {
            console.log('✅ Validation error caught:', error.message);
        }
        
        try {
            validClient.sendData([]);
            console.log('❌ Expected validation error but got success');
        } catch (error) {
            console.log('✅ Empty array validation error caught:', error.message);
        }
        
    } catch (error) {
        console.log('Connection error (expected if service is down):', error.message);
    } finally {
        validClient.stop();
    }
}

/**
 * Main function to run streaming examples
 */
async function main() {
    console.log('Anomaly Detection JavaScript Streaming Examples');
    console.log('='.repeat(50));
    
    const examples = {
        '1': { name: 'Automated Streaming', func: automatedStreamingExample },
        '2': { name: 'Interactive Streaming', func: interactiveStreamingExample },
        '3': { name: 'Batch Streaming', func: batchStreamingExample },
        '4': { name: 'Performance Test', func: performanceStreamingExample },
        '5': { name: 'Error Handling', func: streamingErrorHandlingExample }
    };
    
    // Check if running in interactive mode
    if (process.argv.includes('--interactive')) {
        console.log('\nAvailable examples:');
        Object.entries(examples).forEach(([key, { name }]) => {
            console.log(`  ${key}. ${name}`);
        });
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        
        rl.question('\nSelect example (1-5) or "all" to run all: ', async (choice) => {
            rl.close();
            
            if (choice.toLowerCase() === 'all') {
                for (const [key, { name, func }] of Object.entries(examples)) {
                    console.log(`\n${'='.repeat(20)} ${name} ${'='.repeat(20)}`);
                    try {
                        await func();
                    } catch (error) {
                        console.error('Example failed:', error.message);
                    }
                    
                    if (key !== '5') { // Don't wait after last example
                        console.log('\nPress Enter to continue to next example...');
                        await new Promise(resolve => {
                            const rl2 = readline.createInterface({
                                input: process.stdin,
                                output: process.stdout
                            });
                            rl2.question('', () => {
                                rl2.close();
                                resolve();
                            });
                        });
                    }
                }
            } else if (examples[choice]) {
                const { name, func } = examples[choice];
                console.log(`\n${'='.repeat(20)} ${name} ${'='.repeat(20)}`);
                await func();
            } else {
                console.log('Invalid choice. Running automated example...');
                await automatedStreamingExample();
            }
            
            console.log('\n' + '='.repeat(50));
            console.log('Streaming examples completed!');
        });
    } else {
        // Run automated example by default
        try {
            await automatedStreamingExample();
        } catch (error) {
            console.error('Example execution failed:', error.message);
            
            if (error instanceof StreamingError || error.message.includes('connect')) {
                console.log('\n💡 Make sure the anomaly detection service is running at ws://localhost:8000/ws/stream');
            }
        }
        
        console.log('\n' + '='.repeat(50));
        console.log('Streaming examples completed!');
    }
}

// Run examples if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    AnomalyDetectionStream,
    interactiveStreamingExample,
    automatedStreamingExample,
    batchStreamingExample,
    performanceStreamingExample,
    streamingErrorHandlingExample
};

File: src/packages/data/anomaly_detection/sdk/examples/go/streaming_detection.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"log"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	anomaly "github.com/anomaly-detection/go-sdk"
)

// AnomalyDetectionStream wraps the streaming client with additional functionality
type AnomalyDetectionStream struct {
	client       *anomaly.StreamingClient
	anomalyCount int
	totalPoints  int
	running      bool
	mu           sync.RWMutex
}

// NewAnomalyDetectionStream creates a new streaming wrapper
func NewAnomalyDetectionStream(wsURL string) *AnomalyDetectionStream {
	config := anomaly.StreamingClientConfig{
		WSURL: wsURL,
		Config: anomaly.StreamingConfig{
			BufferSize:         50,
			DetectionThreshold: 0.6,
			BatchSize:          5,
			Algorithm:          anomaly.IsolationForest,
			AutoRetrain:        false,
		},
		AutoReconnect:  true,
		ReconnectDelay: 3 * time.Second,
	}

	client := anomaly.NewStreamingClient(config)

	stream := &AnomalyDetectionStream{
		client: client,
	}

	stream.setupHandlers()
	return stream
}

func (s *AnomalyDetectionStream) setupHandlers() {
	handlers := anomaly.StreamingHandlers{
		OnConnect: func() {
			fmt.Println("✅ Connected to streaming service")
			s.mu.Lock()
			s.running = true
			s.mu.Unlock()
		},
		OnDisconnect: func() {
			fmt.Println("❌ Disconnected from streaming service")
			s.mu.Lock()
			s.running = false
			s.mu.Unlock()
		},
		OnAnomaly: func(anomalyData anomaly.AnomalyData) {
			s.mu.Lock()
			s.anomalyCount++
			count := s.anomalyCount
			s.mu.Unlock()

			fmt.Println("🚨 ANOMALY DETECTED:")
			fmt.Printf("   Index: %d\n", anomalyData.Index)
			fmt.Printf("   Score: %.4f\n", anomalyData.Score)
			fmt.Printf("   Data Point: %v\n", formatFloatSlice(anomalyData.DataPoint, 3))
			if anomalyData.Confidence != nil {
				fmt.Printf("   Confidence: %.4f\n", *anomalyData.Confidence)
			}
			fmt.Printf("   Total anomalies so far: %d\n", count)
			fmt.Println(strings.Repeat("-", 50))
		},
		OnError: func(err error) {
			fmt.Printf("❌ Streaming error: %v\n", err)
		},
		OnMessage: func(data map[string]interface{}) {
			// Handle any other message types if needed
			if msgType, ok := data["type"].(string); ok && msgType != "anomaly" && msgType != "ping" {
				fmt.Printf("📨 Received message: %v\n", data)
			}
		},
	}

	s.client.SetHandlers(handlers)
}

func (s *AnomalyDetectionStream) Start() error {
	fmt.Println("Starting streaming anomaly detection...")
	return s.client.Start()
}

func (s *AnomalyDetectionStream) Stop() {
	fmt.Println("Stopping streaming...")
	s.client.Stop()
	s.mu.Lock()
	s.running = false
	s.mu.Unlock()
}

func (s *AnomalyDetectionStream) SendDataPoint(dataPoint []float64) error {
	err := s.client.SendData(dataPoint)
	if err == nil {
		s.mu.Lock()
		s.totalPoints++
		s.mu.Unlock()
	}
	return err
}

func (s *AnomalyDetectionStream) IsRunning() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.running
}

func (s *AnomalyDetectionStream) GetStats() (int, int) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.totalPoints, s.anomalyCount
}

func (s *AnomalyDetectionStream) GenerateAndSendData(duration time.Duration, interval time.Duration) {
	fmt.Printf("Generating sample data for %v...\n", duration)
	fmt.Println("Normal data will be around [0, 0], anomalies around [5, 5]")
	fmt.Println(strings.Repeat("-", 50))

	rand.Seed(42)
	startTime := time.Now()

	for time.Since(startTime) < duration && s.IsRunning() {
		var dataPoint []float64

		// Generate mostly normal data with occasional anomalies
		if rand.Float64() < 0.1 { // 10% chance of anomaly
			// Anomalous data point
			dataPoint = []float64{
				rand.Float64()*1 + 4.5, // Between 4.5 and 5.5
				rand.Float64()*1 + 4.5,
			}
			fmt.Printf("📤 Sending anomalous point: %v\n", formatFloatSlice(dataPoint, 3))
		} else {
			// Normal data point
			dataPoint = []float64{
				(rand.Float64() - 0.5) * 2, // Between -1 and 1
				(rand.Float64() - 0.5) * 2,
			}
			fmt.Printf("📤 Sending normal point: %v\n", formatFloatSlice(dataPoint, 3))
		}

		if err := s.SendDataPoint(dataPoint); err != nil {
			fmt.Printf("Error sending data: %v\n", err)
		}

		time.Sleep(interval)
	}

	totalPoints, anomalyCount := s.GetStats()
	fmt.Printf("\n✅ Sent %d data points\n", totalPoints)
	fmt.Printf("✅ Detected %d anomalies\n", anomalyCount)
}

// Helper function to format float slices
func formatFloatSlice(slice []float64, precision int) []string {
	result := make([]string, len(slice))
	for i, v := range slice {
		result[i] = fmt.Sprintf("%."+strconv.Itoa(precision)+"f", v)
	}
	return result
}

// interactiveStreamingExample allows user to input data interactively
func interactiveStreamingExample() error {
	fmt.Println("=== Interactive Streaming Example ===")

	stream := NewAnomalyDetectionStream("ws://localhost:8000/ws/stream")
	if err := stream.Start(); err != nil {
		return fmt.Errorf("failed to start streaming: %w", err)
	}
	defer stream.Stop()

	// Wait for connection
	time.Sleep(2 * time.Second)

	if !stream.IsRunning() {
		return fmt.Errorf("failed to connect to streaming service")
	}

	fmt.Println("\nInteractive mode - Enter data points as comma-separated values")
	fmt.Println("Example: 1.5,2.3 or 5.0,5.0 (for anomaly)")
	fmt.Println("Type 'quit' to exit")

	scanner := bufio.NewScanner(os.Stdin)

	for stream.IsRunning() {
		fmt.Print("\nEnter data point (x,y): ")
		if !scanner.Scan() {
			break
		}

		input := strings.TrimSpace(scanner.Text())

		if strings.ToLower(input) == "quit" || strings.ToLower(input) == "exit" || input == "q" {
			break
		}

		// Parse input
		parts := strings.Split(input, ",")
		if len(parts) != 2 {
			fmt.Println("Please enter exactly 2 values separated by comma")
			continue
		}

		var values []float64
		valid := true

		for _, part := range parts {
			val, err := strconv.ParseFloat(strings.TrimSpace(part), 64)
			if err != nil {
				fmt.Println("Invalid input. Please enter numeric values separated by comma")
				valid = false
				break
			}
			values = append(values, val)
		}

		if !valid {
			continue
		}

		if err := stream.SendDataPoint(values); err != nil {
			fmt.Printf("Error sending data: %v\n", err)
		} else {
			fmt.Printf("Sent: %v\n", formatFloatSlice(values, 3))
		}
	}

	totalPoints, anomalyCount := stream.GetStats()
	fmt.Printf("\nSession summary:\n")
	fmt.Printf("Total points sent: %d\n", totalPoints)
	fmt.Printf("Anomalies detected: %d\n", anomalyCount)

	return nil
}

// automatedStreamingExample runs automated data generation
func automatedStreamingExample() error {
	fmt.Println("=== Automated Streaming Example ===")

	stream := NewAnomalyDetectionStream("ws://localhost:8000/ws/stream")
	if err := stream.Start(); err != nil {
		return fmt.Errorf("failed to start streaming: %w", err)
	}
	defer stream.Stop()

	// Wait for connection
	time.Sleep(2 * time.Second)

	if !stream.IsRunning() {
		return fmt.Errorf("failed to connect to streaming service")
	}

	// Run automated data generation
	stream.GenerateAndSendData(20*time.Second, 500*time.Millisecond)

	return nil
}

// batchStreamingExample sends data in batches
func batchStreamingExample() error {
	fmt.Println("=== Batch Streaming Example ===")

	// Generate batch data
	rand.Seed(42)
	var normalBatch [][]float64
	var anomalyBatch [][]float64

	for i := 0; i < 20; i++ {
		point := []float64{
			(rand.Float64() - 0.5) * 2,
			(rand.Float64() - 0.5) * 2,
		}
		normalBatch = append(normalBatch, point)
	}

	for i := 0; i < 5; i++ {
		point := []float64{
			rand.Float64()*1 + 4.5,
			rand.Float64()*1 + 4.5,
		}
		anomalyBatch = append(anomalyBatch, point)
	}

	stream := NewAnomalyDetectionStream("ws://localhost:8000/ws/stream")
	if err := stream.Start(); err != nil {
		return fmt.Errorf("failed to start streaming: %w", err)
	}
	defer stream.Stop()

	// Wait for connection
	time.Sleep(2 * time.Second)

	if !stream.IsRunning() {
		return fmt.Errorf("failed to connect to streaming service")
	}

	// Send normal data batch
	fmt.Println("Sending normal data batch...")
	for i, point := range normalBatch {
		if err := stream.SendDataPoint(point); err != nil {
			fmt.Printf("Error sending data: %v\n", err)
			continue
		}
		fmt.Printf("Sent normal point %d/20: %v\n", i+1, formatFloatSlice(point, 3))
		time.Sleep(200 * time.Millisecond)
	}

	// Send anomalous data batch
	fmt.Println("\nSending anomalous data batch...")
	for i, point := range anomalyBatch {
		if err := stream.SendDataPoint(point); err != nil {
			fmt.Printf("Error sending data: %v\n", err)
			continue
		}
		fmt.Printf("Sent anomaly point %d/5: %v\n", i+1, formatFloatSlice(point, 3))
		time.Sleep(200 * time.Millisecond)
	}

	// Wait for processing
	fmt.Println("\nWaiting for processing to complete...")
	time.Sleep(5 * time.Second)

	return nil
}

// performanceStreamingExample tests streaming performance
func performanceStreamingExample() error {
	fmt.Println("=== Performance Streaming Example ===")

	// Configure for high throughput
	config := anomaly.StreamingClientConfig{
		WSURL: "ws://localhost:8000/ws/stream",
		Config: anomaly.StreamingConfig{
			BufferSize:         100,
			DetectionThreshold: 0.5,
			BatchSize:          20, // Larger batches for efficiency
			Algorithm:          anomaly.IsolationForest,
		},
		AutoReconnect: true,
	}

	client := anomaly.NewStreamingClient(config)

	// Performance tracking
	var (
		anomalyCount int
		startTime    time.Time
		pointsSent   int
		mu           sync.Mutex
	)

	handlers := anomaly.StreamingHandlers{
		OnConnect: func() {
			fmt.Println("✅ Connected - Starting performance test")
			startTime = time.Now()
		},
		OnAnomaly: func(anomalyData anomaly.AnomalyData) {
			mu.Lock()
			anomalyCount++
			currentCount := anomalyCount
			currentPoints := pointsSent
			mu.Unlock()

			if currentCount%10 == 0 {
				elapsed := time.Since(startTime).Seconds()
				rate := float64(currentPoints) / elapsed
				fmt.Printf("📊 Performance: %d anomalies detected, %.1f points/sec\n",
					currentCount, rate)
			}
		},
		OnError: func(err error) {
			fmt.Printf("❌ Error: %v\n", err)
		},
	}

	client.SetHandlers(handlers)

	if err := client.Start(); err != nil {
		return fmt.Errorf("failed to start streaming: %w", err)
	}
	defer client.Stop()

	// Wait for connection
	time.Sleep(2 * time.Second)

	// Generate high volume of data
	rand.Seed(42)
	fmt.Println("Sending high volume data stream (1000 points)...")

	for i := 0; i < 1000; i++ {
		if i%100 == 0 {
			fmt.Printf("Sent %d/1000 points...\n", i)
		}

		// Mix of normal and anomalous data
		var dataPoint []float64
		if rand.Float64() < 0.05 { // 5% anomalies
			dataPoint = []float64{
				rand.Float64()*1 + 3.5,
				rand.Float64()*1 + 3.5,
			}
		} else {
			dataPoint = []float64{
				(rand.Float64() - 0.5) * 2,
				(rand.Float64() - 0.5) * 2,
			}
		}

		if err := client.SendData(dataPoint); err != nil {
			fmt.Printf("Error sending data: %v\n", err)
			continue
		}

		mu.Lock()
		pointsSent++
		mu.Unlock()

		// Small delay to avoid overwhelming
		time.Sleep(10 * time.Millisecond)
	}

	// Wait for final processing
	fmt.Println("Waiting for final processing...")
	time.Sleep(5 * time.Second)

	// Calculate final statistics
	totalTime := time.Since(startTime).Seconds()
	throughput := float64(pointsSent) / totalTime

	fmt.Println("\n📊 Performance Results:")
	fmt.Printf("   Total points: %d\n", pointsSent)
	fmt.Printf("   Total time: %.2f seconds\n", totalTime)
	fmt.Printf("   Throughput: %.1f points/second\n", throughput)
	fmt.Printf("   Anomalies detected: %d\n", anomalyCount)
	fmt.Printf("   Anomaly rate: %.2f%%\n", float64(anomalyCount)/float64(pointsSent)*100)

	return nil
}

// streamingErrorHandlingExample demonstrates error handling
func streamingErrorHandlingExample() error {
	fmt.Println("=== Streaming Error Handling Example ===")

	// Test with invalid URL
	invalidConfig := anomaly.StreamingClientConfig{
		WSURL:         "ws://invalid-url:9999/ws/stream",
		AutoReconnect: false,
	}

	invalidClient := anomaly.NewStreamingClient(invalidConfig)
	errorReceived := false

	handlers := anomaly.StreamingHandlers{
		OnError: func(err error) {
			fmt.Printf("✅ Expected connection error: %v\n", err)
			errorReceived = true
		},
	}

	invalidClient.SetHandlers(handlers)

	if err := invalidClient.Start(); err != nil {
		fmt.Printf("✅ Connection failed as expected: %v\n", err)
	} else {
		// Wait a bit to see if error is received
		time.Sleep(5 * time.Second)
		if !errorReceived {
			fmt.Println("❌ Expected error but none received")
		}
	}

	invalidClient.Stop()

	// Test sending invalid data
	fmt.Println("\nTesting invalid data handling...")
	validConfig := anomaly.StreamingClientConfig{
		WSURL: "ws://localhost:8000/ws/stream",
	}

	validClient := anomaly.NewStreamingClient(validConfig)

	if err := validClient.Start(); err != nil {
		fmt.Printf("Connection error (expected if service is down): %v\n", err)
		return nil
	}
	defer validClient.Stop()

	time.Sleep(2 * time.Second)

	// Try to send invalid data
	fmt.Println("Testing empty data point...")
	if err := validClient.SendData([]float64{}); err != nil {
		fmt.Printf("✅ Empty array validation error caught: %v\n", err)
	} else {
		fmt.Println("❌ Expected validation error but got success")
	}

	return nil
}

func main() {
	fmt.Println("Anomaly Detection Go Streaming Examples")
	fmt.Println(strings.Repeat("=", 50))

	examples := map[string]struct {
		name string
		fn   func() error
	}{
		"1": {"Automated Streaming", automatedStreamingExample},
		"2": {"Interactive Streaming", interactiveStreamingExample},
		"3": {"Batch Streaming", batchStreamingExample},
		"4": {"Performance Test", performanceStreamingExample},
		"5": {"Error Handling", streamingErrorHandlingExample},
	}

	// Simple command line interface
	if len(os.Args) > 1 {
		choice := os.Args[1]
		if choice == "all" {
			// Run all examples
			for i := 1; i <= 5; i++ {
				key := strconv.Itoa(i)
				if example, exists := examples[key]; exists {
					fmt.Printf("\n%s %s %s\n",
						strings.Repeat("=", 20), example.name, strings.Repeat("=", 20))

					if err := example.fn(); err != nil {
						log.Printf("❌ Example failed: %v", err)
					}

					if i < 5 { // Don't wait after last example
						fmt.Println("\nPress Enter to continue to next example...")
						bufio.NewScanner(os.Stdin).Scan()
					}
				}
			}
		} else if example, exists := examples[choice]; exists {
			fmt.Printf("\n%s %s %s\n",
				strings.Repeat("=", 20), example.name, strings.Repeat("=", 20))
			if err := example.fn(); err != nil {
				log.Printf("❌ Example failed: %v", err)
			}
		} else {
			fmt.Printf("Invalid choice '%s'. Running automated example...\n", choice)
			if err := automatedStreamingExample(); err != nil {
				log.Printf("❌ Example failed: %v", err)
			}
		}
	} else {
		// Show menu
		fmt.Println("\nAvailable examples:")
		for key, example := range examples {
			fmt.Printf("  %s. %s\n", key, example.name)
		}
		fmt.Println("\nUsage:")
		fmt.Println("  go run streaming_detection.go [1-5|all]")
		fmt.Println("  Or run without arguments to see this menu")
		fmt.Println("\nRunning automated example...")

		if err := automatedStreamingExample(); err != nil {
			log.Printf("❌ Example failed: %v", err)

			if _, ok := err.(*anomaly.StreamingError); ok {
				fmt.Println("\n💡 Make sure the anomaly detection service is running at ws://localhost:8000/ws/stream")
			}
		}
	}

	fmt.Println("\n" + strings.Repeat("=", 50))
	fmt.Println("Streaming examples completed!")
}